contracts/contracts/strategies/balancer/BaseBalancerStrategy.sol

```diff
@@ -0,0 +1,457 @@
+        ) / tokens.length);
+
+        /* If pool asset is equals _asset it means a rate provider for that asset
+         * exists and that asset is not necessarily pegged to a unit (ETH).
+         *
+         * Because this function returns the balance of the asset not denominated in
+         * ETH units we need to convert the amount to asset amount.
+         */
+        if (toPoolAsset(_asset) == _asset) {
+            amount = amount.divPrecisely(getRateProviderRate(_asset));
+        }
+    }
+
+    /**
+     * @notice Returns the value of all assets managed by this strategy.
+     * Uses the Balancer pool's rate (virtual price) to convert the strategy's
+     * Balancer Pool Tokens (BPT) to ETH value.
+     * @return value The ETH value
+     *
+     * IMPORTANT if this function is overridden it needs to have a whenNotInVaultContext
+     * modifier on it or it is susceptible to read-only re-entrancy attack
+     */
+    function checkBalance()
+        external
+        view
+        virtual
+        whenNotInVaultContext
+        returns (uint256 value)
+    {
+        uint256 bptBalance = _getBalancerPoolTokens();
+
+        // Convert BPT to ETH value
+        value = bptBalance.mulTruncate(
+            IRateProvider(platformAddress).getRate()
+        );
+    }
+
+    /// @notice Balancer Pool Tokens (BPT) in the Balancer pool.
+    function _getBalancerPoolTokens()
+        internal
+        view
+        virtual
+        returns (uint256 balancerPoolTokens)
+    {
+        balancerPoolTokens = IERC20(platformAddress).balanceOf(address(this));
+    }
+
+    /* solhint-disable max-line-length */
+    /**
+     * @notice BPT price is calculated by dividing the pool (sometimes wrapped) market price by the
+     * rateProviderRate of that asset. To get BPT expected we need to multiply that by underlying
+     * asset amount divided by BPT token rate. BPT token rate is similar to Curve's virtual_price
+     * and expresses how much has the price of BPT appreciated in relation to the underlying assets.
+     *
+     * @dev
+     * bptPrice = pool_asset_oracle_price / pool_asset_rate
+     *
+     * Since we only have oracle prices for the unwrapped version of the assets the equation
+     * turns into:
+     *
+     * bptPrice = from_pool_token(asset_amount).amount * oracle_price / pool_asset_rate
+     *
+     * bptExpected = bptPrice(in relation to specified asset) * asset_amount / BPT_token_rate
+     *
+     * and since from_pool_token(asset_amount).amount and pool_asset_rate cancel each-other out
+     * this makes the final equation:
+     *
+     * bptExpected = oracle_price * asset_amount / BPT_token_rate
+     *
+     * more explanation here:
+     * https://www.notion.so/originprotocol/Support-Balancer-OETH-strategy-9becdea132704e588782a919d7d471eb?pvs=4#382834f9815e46a7937f3acca0f637c5
+     */
+    /* solhint-enable max-line-length */
+    function getBPTExpected(address _asset, uint256 _amount)
```

Comment:
Just a comment, no change required. Our current oracles are fairly bad in terms of latency and accuracy. The default value set for reverting a deposit/withdraw 1e15, or 0.1% slippage + fee difference from an oracle. That feels like it's going to trigger some times. At a minimum during fork tests.

It may not be much of a production problem, because the strategist can always alter the slippage if needed.

Reply:
I have been thinking about this, I think it is problematic in both directions. A lot of ETH pegged oracles have a 24 hour heartbeat and 2% price change trigger. In case e.g. rETH price plunges for 1.5% (rETH -> ETH price that is) and oracles still have not updated, such a low BPTExpected would be set that we could get MEW-ed. 

Check balance wouldn't catch that since that only works off of rate providers. So it wouldn't be shielded with vault value checker.

**Proposed solution:** 
Similar to VaultValue checker we could write an Oracle price checker. This could be an offline brownie module that supplements our allocations scripts and pulls prices from 1inch and passes them to onChain oracle price checker function. If the deviation is too big: 
- in direction of asset price being too high. It could propose to increase the slippage when withdrawing. And prevent deposits since the strategy would be asking for too little BPT
- in direction of asset price being too low. It could propose to increase slippage when depositing.(and undo after deposit) And prevent withdrawing since we'd be leaving too much BPT on the table.

In case we ever make this a default strategy such solutions are of no use of course.